### DISCUSSION TITLE
Marker clusterer with big data is very slow

### DISCUSSION THREAD

**Discussion by leo3620** [visgl/react-google-maps#526] [Q&A]
Hello, I tried to implement the same exemple of clusterer as you provide but with 10k data, but it's very slow and freeze sometimes. Did you try your exemple with that amount of data ? 
I used to make it work some vanilla JS and refs but I wanted to update this to use this library but for now I can't as the performance seems slower than using without lib.

**Comment by usefulthink** [comment]
Which of the examples are you referring to? I actually tested the [Custom Marker Clustering](https://visgl.github.io/react-google-maps/examples/custom-marker-clustering) with over 200k points, without significant problems (besides the raw dataset being too big for an example on our site). But at that point I would usually implement the clustering in a worker, which would further complicate the example.

The other variant where all markers are rendered in React first is quite a bit slower and very sensitive to re-rendering problems.

**Reply by leo3620** [reply]
Thank you for your answer, Indeed I tried with the other variant "[Marker](https://visgl.github.io/react-google-maps/examples/marker-clustering)", because I wanted to use only @googlemaps/markerclusterer as I used before. So for that exemple I tried with an API that return 7.5k data. I clone your repo and I just update the loadTreeDataset function in the exemple. What happens is that it takes 17secondes to load the map I don't know why. I have a blank screen and after that, everything appears. Do you have any idea ? 

Today I tried also with the Custom Marker Clustering as you mentionned, and it seems to work even if it's still a but slow, I mean, it's not ultra smooth but I'm skeptical about using an external lib from MapBox for Clustering while this lib is about google maps tools. And google provides it's own algorithm.

**Reply by usefulthink** [reply]
Don't be skeptical :) The default configuration of the @googlemaps/markerclusterer [uses the very same `supercluster` algorithm](https://github.com/googlemaps/js-markerclusterer/blob/main/src/algorithms/supercluster.ts#L18), which is just the fastest implementation of this algorithm in js. The part that takes some time when the data is updated is that supercluster has to rebuild all the indices / rtrees for every zoom-level – this is where the actual clustering happens. Since that is a compute-intensive task, the only way to get around that is to move clustering into a worker.
